### PR TITLE
Rework linkify-urls-in-code

### LINKED ISSUE #463: URL linkification fails inside a Travis badge
It results in:

```markdown
[![Build Status](https://travis-ci.org/sindresorhus/caprine.svg?branch=master" target="_blank" class="rg-linkified-code">https://travis-ci.org/sindresorhus/caprine.svg?branch=master)](https://travis-ci.org/sindresorhus/caprine)
```

From: https://github.com/sindresorhus/caprine/pull/220/files

### LINKED ISSUE #478: Weird escaping in issue titles with single-quote
screen shot 2017-06-06 at 23 19 36

Example: https://github.com/sindresorhus/refined-github/issues/477

### PR DESCRIPTION
**Pull request by bfred-it**
First commit was to fix #463: If a line contained the same URL twice, it would replace the first URL twice instead of touching the second one.

Then I extended the changes to reorganize the code and test more of it.

Closes #463 
Closes #478

### CODE REVIEW

**Comment by bfred-it** [comment]
I'll reopen this once the build is in, to avoid expensive conflicts.

⬇️❌ Don't delete branch!

**Comment by hkdobrev** [comment]
@bfred-it BTW every branch which participated in a PR could be restored after 😉

**Comment by bfred-it** [comment]
WIP. I'm extending this

**Comment by bfred-it** [comment]
These changes can be tested in
* Code: https://github.com/sindresorhus/refined-github/blob/be756e68dd4bf24d53e802e1fec126506979fa80/test/linkify-urls-in-code.js
* Code with duplicate URLs: https://github.com/sindresorhus/caprine/pull/220/files
* Title: https://github.com/sindresorhus/refined-github/pull/299

Let me know how the tests look

**Comment by sindresorhus** [comment]
Could you use https://github.com/sindresorhus/linkify-issues and https://github.com/sindresorhus/linkify-urls while you're at it?

**Comment by bfred-it** [comment]
@sindresorhus damn you, you could have done that 48 hours ago :D

**Comment by bfred-it** [comment]
I think `linkify-urls` has issues here: https://github.com/sindresorhus/caprine/pull/220/files

It only detects a single URL:

```
https://travis-ci.org/sindresorhus/caprine.svg?branch=master)](https://travis-ci.org/sindresorhus/caprine
```

**Comment by sindresorhus** [comment]
> damn you, you could have done that 48 hours ago :D

I was actually going to do a PR adding those after your "build script" PR, but you're just too fast. (Not complaining 😛)

**Comment by bfred-it** [comment]
What do we do about `url-regex`?

**Comment by sindresorhus** [comment]
@bfred-it I've worked around it for now: https://github.com/sindresorhus/linkify-urls/commit/aca71b8855c0b13286308119e427184b8158cbed

**Comment by bfred-it** [comment]
Great! I'll fix this today. I also found another issue that has to do with applying this to HTML rather than its text nodes. e.g. `http://a.com<span` is wholly matched as a URL. I'll use an [DOM tree walker](https://stackoverflow.com/a/10730777/288906)

**Comment by bfred-it** [comment]
It was appearing [here](https://github.com/binarykitchen/videomail-client/blob/feature/rollupjs/package.json) but with `linkify-urls@1.0.2` the issue was fixed.

screen shot 2017-06-13 at 15 27 23

File: src/libs/domify.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// Get DOM node from HTML
+export default html => {
+	// Shortcut for html`text` instead of html(`text`)
+	if (html.raw) {
+		html = String.raw(...arguments);
+	}
```
Comment (sindresorhus) [inline src/libs/domify.js]: I don't like using tagged templates just for the sake of it. In this case it gives no added benefit over a normal function call, right?

File: src/libs/get-text-nodes.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+	const next = () => {
+		const value = walker.nextNode();
+		return {
+			value,
+			done: !value
+		};
+	};
+	walker[Symbol.iterator] = () => ({next});
+	return walker;
+};
```
Comment (sindresorhus) [inline src/libs/get-text-nodes.js]: This would be a useful npm package. Hint hint ;)

File: src/libs/linkify-urls-in-code.js [hunk 3]
```diff
@@ -1,38 +1,57 @@
...
+		editTextNodes(linkifyUrls, el);
+	}
+
+	// Linkify issue refs in comments
+	for (const el of select.all('.blob-code-inner span.pl-c', untouchedCode)) {
+		editTextNodes(linkifyIssues, el);
+	}
+
+	// Mark code block as touched
+	untouchedCode.forEach(el => el.classList.add(linkifiedURLClass));
```
Comment (sindresorhus) [inline src/libs/linkify-urls-in-code.js]: Use a for-of loop for consistency and performance.

File: src/libs/util.js [hunk 4]
```diff
@@ -1,9 +1,4 @@
...
-		const issueParts = issue.split('#');
-		return `<a href="/${issueParts[0]}/issues/${issueParts[1]}" ${attrs}>${issue}</a>`;
-	}
-	return `<a href="/${repoPath}/issues/${issue.replace('#', '')}" ${attrs}>${issue}</a>`;
-};
-
+// Export as functions because of the g flag
+// https://stackoverflow.com/questions/1520800/why-regexp-with-global-flag-in-javascript-give-wrong-results
+export const getURLRegex = () => /(http(s)?(:\/\/))(www\.)?[a-zA-Z0-9-_.]+(\.[a-zA-Z0-9]{2,})([-a-zA-Z0-9:%_+.~#?&//=]*)/g;
+export const getIssueRegex = () => /([a-zA-Z0-9-_.]+\/[a-zA-Z0-9-_.]+)?#[0-9]+/g;
```
Comment (sindresorhus) [inline src/libs/util.js]: Shouldn't these be removed now?

File: src/libs/linkify-urls-in-code.js [hunk 5]
```diff
@@ -24,6 +24,9 @@ export const editTextNodes = (fn, el) => {
 		return;
 	}
 	for (const textNode of getTextNodes(el)) {
+		if (textNode.textContent.length < 11) { // Shortest url: http://j.mp
```
Comment (sindresorhus) [inline src/libs/linkify-urls-in-code.js]: Clever 🙌

File: src/libs/get-text-nodes.js (hunk 2 above)
Comment (bfred-it) [inline src/libs/get-text-nodes.js]: I thought about t but I'm not good at testing browser packages 😅

File: src/libs/util.js (hunk 4 above)
Comment (bfred-it) [inline src/libs/util.js]: I suppose so. The rebase conflicts were confusing

File: src/libs/linkify-urls-in-code.js (hunk 5 above)
Comment (bfred-it) [inline src/libs/linkify-urls-in-code.js]: Realistically we can even increase the number to skip already-short URLs like that.

File: src/libs/get-text-nodes.js (hunk 2 above)
Comment (sindresorhus) [inline src/libs/get-text-nodes.js]: Yeah, it's annoying, but easier if JSDom supports the tree-walker API. See: https://github.com/sindresorhus/is-blob/blob/927899acea40ce7bd31e2b3c4352248eba3a0609/test.js

**Comment by bfred-it** [comment]
Good to go?

**Comment by sindresorhus** [comment]
Yup. It's perfect now.

**Pull request merged**